File: benches/phased_reuse.py
"""Phased-reuse figure: the analyze-once / factor-many amortization.

For each matrix the `phased_reuse` bench records the analyze and factor times.
A frequency sweep / Newton iteration factors K value sets that share one
sparsity pattern, so the cost is:

* analyze-once (RSLAB phased): `analyze + K * factor`
* analyze-each (naive):        `K * (analyze + factor)`

This plots cumulative wall-clock vs K for both, per matrix - the gap is the
amortized analysis, the reason RSLAB exposes a reusable `LdltSymbolic`.

Run:  python benches/phased_reuse.py [phased_reuse.jsonl]
"""
import json
import sys
from pathlib import Path

import numpy as np
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt

import bench_style
from bench_style import GRAY, BLUE_SHADES

COLORS = BLUE_SHADES


def main():
    path = Path(sys.argv[1]) if len(sys.argv) > 1 else Path("benches/bench_out/phased_reuse.jsonl")
    recs = [json.loads(l) for l in open(path) if l.strip()]
    # Three representative matrices spanning the analyze fraction (low/mid/high).
    recs.sort(key=lambda r: r["analyze_ms"] / (r["analyze_ms"] + r["factor_ms"]))
    sel = [recs[0], recs[len(recs) // 2], recs[-1]] if len(recs) >= 3 else recs

    bench_style.setup()
    ks = np.arange(1, 21)
    fig, ax = plt.subplots(figsize=(8.5, 6.4))
    for r, c in zip(sel, COLORS):
        a, f = r["analyze_ms"], r["factor_ms"]
        # Speedup of reusing the analysis vs re-analyzing each of K factorizations:
        # K*(a+f) / (a + K*f), rising from 1 to the asymptote 1 + a/f.
        speedup = ks * (a + f) / (a + ks * f)
        asymp = (a + f) / f
        ax.plot(ks, speedup, color=c, lw=2.4, marker="o", ms=4,
                label=f"{r['name']}  (analyze {100*a/(a+f):.0f}%  →  max {asymp:.2f}x)")
        ax.axhline(asymp, color=c, ls=":", lw=1.2, alpha=0.5)
    ax.axhline(1.0, color=GRAY, lw=1, alpha=0.5)
    ax.set_xlabel("factorizations sharing the pattern  (K, e.g. frequency points)")
    ax.set_ylabel("speedup: re-analyze each  /  analyze once")
    ax.set_title("Phased reuse: speedup from reusing the analysis over K factorizations")
    ax.set_xlim(1, ks[-1])
    ax.grid(True, ls=":", alpha=0.4)
    out = path.parent / "phased_reuse.png"
    bench_style.legend_below(fig, ax=ax)
    fig.savefig(out, dpi=150, transparent=True, bbox_inches="tight")
    print(f"wrote {out}")
    for r in recs:
        a, f = r["analyze_ms"], r["factor_ms"]
        sp = (a + f) / f  # asymptotic speedup of phased vs naive (K -> inf)
        print(f"  n={r['n']:>7}  analyze {a:>8.2f}ms  factor {f:>9.2f}ms  "
              f"asymptotic phased speedup x{sp:.2f}")


if __name__ == "__main__":
    main()

File: benches/phased_reuse.rs
//! Phased-reuse benchmark: the PARDISO-style "analyze once, factor many" workflow
//! that a frequency sweep / Newton iteration exercises (same sparsity pattern,
//! changing values). Measures the analyze phase (fill-reducing ordering +
//! symbolic, value-independent) and the numeric factor separately per matrix, so
//! `phased_reuse.py` can show the amortized cost over K factorizations:
//! analyze-once (`analyze + K*factor`) vs analyze-each (`K*(analyze + factor)`).
//!
//! Helmholtz 3D matrices stand in for an EM/FEM frequency sweep: one pattern,
//! many `k²`. Correctness of reuse is asserted (the analysis factors a second
//! value set and still solves).
//!
//! Run: `cargo bench --bench phased_reuse --features matgen`
#![allow(clippy::needless_range_loop)]

use std::io::Write;
use std::time::Instant;

use num_complex::Complex;
use rslab::matgen::stencil;
use rslab::{CscMatrix, SolverSettings, LdltSymbolic};

type C = Complex<f64>;

fn best_of<F: FnMut() -> f64>(reps: usize, mut f: F) -> f64 {
    (0..reps).map(|_| f()).fold(f64::INFINITY, f64::min)
}

fn main() {
    let out_path =
        std::env::var("RLA_BENCH_OUT").unwrap_or_else(|_| "benches/bench_out/phased_reuse.jsonl".to_string());
    if let Some(d) = std::path::Path::new(&out_path).parent() {
        let _ = std::fs::create_dir_all(d);
    }
    let mut out = std::fs::File::create(&out_path).expect("open out");
    let opts = SolverSettings::default();
    let so = stencil::StencilOpts::default();

    // A spread of fill profiles: the analyze/factor ratio (hence the reuse win)
    // grows as the factor gets cheaper relative to the value-independent analysis.
    // Helmholtz 3D (high fill, factor-dominated) is the EM frequency-sweep case;
    // 2D Poisson and banded (low fill) show where reuse pays most.
    let mut cases: Vec<(String, CscMatrix<C>, CscMatrix<C>)> = Vec::new();
    for k in [20usize, 28, 36] {
        cases.push((
            format!("helmholtz3d_{}", k * k * k),
            stencil::helmholtz(&[k, k, k], Complex::new(2.0, 0.1), &so),
            stencil::helmholtz(&[k, k, k], Complex::new(3.5, 0.2), &so),
        ));
    }
    for m in [200usize, 360] {
        cases.push((
            format!("helmholtz2d_{}", m * m),
            stencil::helmholtz(&[m, m], Complex::new(0.5, 0.05), &so),
            stencil::helmholtz(&[m, m], Complex::new(0.9, 0.1), &so),
        ));
    }
    for n in [40000usize, 80000] {
        cases.push((
            format!("banded_{n}"),
            rslab::matgen::structured::banded::<C>(n, 24, 1.0, 1),
            rslab::matgen::structured::banded::<C>(n, 24, 1.3, 2),
        ));
    }

    for (name, a, a2) in &cases {
        let n = a.n;
        let nnz = a.values.len();

        let analyze_ms = best_of(3, || {
            let t = Instant::now();
            let s = LdltSymbolic::analyze(a).expect("analyze");
            let ms = t.elapsed().as_secs_f64() * 1e3;
            std::hint::black_box(s.n());
            ms
        });
        // Reuse one analysis for the factor timing (the phased path).
        let sym = LdltSymbolic::analyze(a).expect("analyze");
        let factor_ms = best_of(3, || {
            let t = Instant::now();
            let f = sym.factor(a, &opts).expect("factor");
            let ms = t.elapsed().as_secs_f64() * 1e3;
            std::hint::black_box(f.factor_nnz());
            ms
        });

        // Reuse correctness: factor a *second* value set on the same analysis.
        let f2 = sym.factor(a2, &opts).expect("factor a2");
        let b: Vec<C> = (0..n).map(|i| Complex::new(i as f64 % 7.0 - 3.0, 1.0)).collect();
        let x = f2.solve(&b).expect("solve");
        let mut ax = vec![C::new(0.0, 0.0); n];
        a2.symv(&x, &mut ax);
        let res = (0..n).map(|i| (ax[i] - b[i]).norm_sqr()).sum::<f64>().sqrt()
            / b.iter().map(|v| v.norm_sqr()).sum::<f64>().sqrt();
        assert!(res < 1e-8, "reuse solve residual {res:.1e}");

        let rec = format!(
            "{{\"name\":\"{name}\",\"n\":{n},\"nnz\":{nnz},\"analyze_ms\":{analyze_ms:.3},\"factor_ms\":{factor_ms:.3}}}"
        );
        writeln!(out, "{rec}").expect("write");
        eprintln!(
            "[phased] {name:<16} n={n:>7} nnz={nnz:>9}  analyze {analyze_ms:>8.2}ms  factor {factor_ms:>9.2}ms  (analyze {:.0}% of one solve)",
            100.0 * analyze_ms / (analyze_ms + factor_ms)
        );
    }
    eprintln!("[phased] wrote {out_path}");
}
